## task
- format : pred out > train
- data : pred + train 
- merge tags (BII)
- convert to re input

In [1]:
import pandas as pd
import csv

## pred_out.txt 를 train.tsv 의 형식으로 바꾸기

In [29]:
pred_sent_list = []
pred_tag_list = []

pred_list = [line.rstrip() for line in open('KoBERT-NER/sample_pred_out.txt', encoding = 'utf8')]
for sample in pred_list:
    sample_split = [word.replace('[', '').replace(']','').replace(':','') for word in sample.split(' ')]
    labels = ['UNK','VACC-B','VACC-I','SYMP-B','SYMP-I','DIS-B','DIS-I']
    word_list = []
    label_list = []
    for word in sample_split :
        if word.endswith(tuple(labels)):
            for label in labels:
                if word.endswith(label):
                    word_list.append(word.replace(label,''))
                    label_list.append(label) 
        else:
            label_list.append('O')
            word_list.append(word)

    pred_sent_list.append([' '.join(word_list)])
    pred_tag_list.append([' '.join(label_list)])

In [30]:
print(len(pred_sent_list))
print(pred_sent_list[:10])
print('----')
print(len(pred_tag_list))
print(pred_tag_list[:10])

1435
[['이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 통증, 부기 ▲발열, 메스꺼움, 근육통 ▲피로감 ▲두통 등을 안내하고 있으나 생리불순이나'], ['앞서 해외에서도 백신 접종 초기부터 생리 주기가 뒤바뀌거나 생리 양이 눈에 띄게 늘거나 주는 등 이상 반응을 겪었다는 여성들의 증언이 잇따랐다 .'], ['이에 미국 일리노이대 연구진은 백신 접종 이후 생리 이상 증상을 겪은 여성 14만 여명의 사례를 모아 보고서를 작성했다 .'], ['사례 중에는 몇년간 생리를 하지 않았다가 백신 접종 후 출혈이 생겼다는 여성도 있었다 .'], ['이후 미국 질병통제예방센터(CDC)는 여성들의 생리 반응에 대한 자체 조사에 나섰고, 최근 여성의 생리불순 등을 잠재적인 부작용으로 확인했다 .'], ['다만 현지 의사들은 생리불순 등의 부작용은 걱정할 정도가 아니며, 백신을 맞지 않을 이유는 되지 않는다고 설명했다 .'], ['정부가 지난 16일 혈소판감소성 혈전증 확정 판정을 받은 30대 사망자에 대해 코로나19(COVID-19) 백신 접종과의 인과성을 인정했다 .'], ['이 환자의 유족들이 보상을 신청할 경우 예방접종피해보상 전문위원회 보상심의를 거쳐 사망자일시보상금 및 장제비를 받을 수 있다 .'], ['코로나19 예방접종피해 국가보상제도에서 책정한 최대 금액은 사망일시보상금 4억3740만원이다 .'], ['21일 코로나19 예방접종피해조사반은 18차 예방접종피해조사반 회의를 통해 지난 16일 사망한 국내 두번째 혈소판감소성 혈전증 확정사례 환자 사례의 인과성을']]
----
1435
[['O O O O O O O O O SYMP-B SYMP-B SYMP-B SYMP-B SYMP-B SYMP-B SYMP-B O O O SYMP-B'], ['O O O O O O O O O O O O O O O O O O O O O O'], ['O O O O O O O SYMP-B SYMP-I O O O O O O O O O O'], ['O O O O O O O O 

## train.tsv 를 읽어서 이중 리스트 형식으로 변경

In [31]:
train_sent_list = []
train_tag_list = []
with open('KoBERT-NER/data/train.tsv', encoding = 'utf8') as file:
    tsv_file = csv.reader(file, delimiter="\t")
    for line in tsv_file:
        train_sent_list.append([line[0]])
        train_tag_list.append([line[1]])

In [32]:
print(len(train_sent_list))
print(train_sent_list[:10])
print('----')
print(len(train_tag_list))
print(train_tag_list[:10])

1582
[['급성기의 심근염에서 회복된 이후에도 심부전 증세가 간헐적으로 나타나기도 하고, 드물게 재발이 되는 경우도 있다 .'], ['그러나 방귀가 사라지지 않거나 다른 소화기증세와 함께 곤란을 겪는다면 가정의학과나 내과 전문의를 찾아가는 것이 좋다 .'], ['또 다른 미접종자는 어린이집 등원 차 기다리는데 화이자 1차 접종했다던 아이 친구 엄마랑 잠깐 이야기 나누는데 병원 소독약 냄새가 났다 .'], ['당장 지난 9일부터 시작된 만 18~49세의 예방접종 사전예약률은 전체 목표치인 70%보다 미달(14일 오후 6시 기준 58.1%)인 상태다 .'], ['병원에서 뇌 관련 검사만 받다가 코로나 관련 의심이 있다고 알려진 심장 쪽 검사는 받지도 못하고, 치료다운 치료는 받아 보지도 못하고 8일 오전 3시 41분경에 사망했다”고 적었다 .'], ['먼저 이날 질병청이 공개한 예방접종피해조사반이 중증 이상반응을 심의해 인과성을 인정한 사례는 총 303건으로 중증이상반응 신고 총 2440건의 12.4%에 불과했는데, 이에 대해 강기윤 국민의힘 의원은 국민들 입장에서는 이상사례들이 인정받지 못하니까 백신접종에 대한 불신이 생기는 것이라면서 백신접종 이후 정부차원에서 국민들이 유급휴가를 사용해 푹 쉴 수 있도록 도와주는 등 대책을 마련해야한다고 꼬집었다 .'], ['◇지난해 독감 발생 적었는데… 지난해에는 많은 사람들이 마스크 착용과 개인 위생 수칙을 철저히 지킴에 따라 독감 발생이 크게 적었다 .'], ['접종 후 발열, 근육통이 심하다면 진통해열제를 복용하는 것이 증상을 개선하는 데 도움이 될 수 있고, 접종 후 가능하면 충분한 휴식을 취하는 것이 바람직하겠습니다 .'], ['뇌출혈, 뇌경색, 급성림프구성백혈병 등으로 사망한 경우도 눈에 띈다 .'], ['진료비는 본인이 부담한 금액에 전액 지급되고 간병비는 일 5만원, 장제비는 30만원이 지급된다 .']]
----
1582
[['O O O O O O O O O O O O O O O'], ['O O O O O 

## pred, train data 합치기
### sent_list, tag_list
- 1435 + 1582 = 3017 확인 완료

In [33]:
len(pred_sent_list) + len(train_sent_list)

3017

In [34]:
sent_list = pred_sent_list + train_sent_list
tag_list = pred_tag_list + train_tag_list

In [35]:
print(len(sent_list))
print(sent_list[:10])
print(len(tag_list))
print(tag_list[:10])

3017
[['이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 통증, 부기 ▲발열, 메스꺼움, 근육통 ▲피로감 ▲두통 등을 안내하고 있으나 생리불순이나'], ['앞서 해외에서도 백신 접종 초기부터 생리 주기가 뒤바뀌거나 생리 양이 눈에 띄게 늘거나 주는 등 이상 반응을 겪었다는 여성들의 증언이 잇따랐다 .'], ['이에 미국 일리노이대 연구진은 백신 접종 이후 생리 이상 증상을 겪은 여성 14만 여명의 사례를 모아 보고서를 작성했다 .'], ['사례 중에는 몇년간 생리를 하지 않았다가 백신 접종 후 출혈이 생겼다는 여성도 있었다 .'], ['이후 미국 질병통제예방센터(CDC)는 여성들의 생리 반응에 대한 자체 조사에 나섰고, 최근 여성의 생리불순 등을 잠재적인 부작용으로 확인했다 .'], ['다만 현지 의사들은 생리불순 등의 부작용은 걱정할 정도가 아니며, 백신을 맞지 않을 이유는 되지 않는다고 설명했다 .'], ['정부가 지난 16일 혈소판감소성 혈전증 확정 판정을 받은 30대 사망자에 대해 코로나19(COVID-19) 백신 접종과의 인과성을 인정했다 .'], ['이 환자의 유족들이 보상을 신청할 경우 예방접종피해보상 전문위원회 보상심의를 거쳐 사망자일시보상금 및 장제비를 받을 수 있다 .'], ['코로나19 예방접종피해 국가보상제도에서 책정한 최대 금액은 사망일시보상금 4억3740만원이다 .'], ['21일 코로나19 예방접종피해조사반은 18차 예방접종피해조사반 회의를 통해 지난 16일 사망한 국내 두번째 혈소판감소성 혈전증 확정사례 환자 사례의 인과성을']]
3017
[['O O O O O O O O O SYMP-B SYMP-B SYMP-B SYMP-B SYMP-B SYMP-B SYMP-B O O O SYMP-B'], ['O O O O O O O O O O O O O O O O O O O O O O'], ['O O O O O O O SYMP-B SYMP-I O O O O O O O O O O'], ['O O O O O O O O O SYM

In [36]:
sent_list[0][0].split()

['이런',
 '가운데',
 '질병관리청은',
 '예방',
 '접종',
 '후',
 '이상',
 '반응으로',
 '▲접종부위',
 '통증,',
 '부기',
 '▲발열,',
 '메스꺼움,',
 '근육통',
 '▲피로감',
 '▲두통',
 '등을',
 '안내하고',
 '있으나',
 '생리불순이나']

In [42]:
new_sent_list=[]
for sent_one in sent_list:
    new_sent_list.append(sent_one[0].split())

new_tag_list=[]
for tag_one in tag_list:
    new_tag_list.append(tag_one[0].split())

# merge tags

In [44]:
from itertools import combinations

def merge_tags(tags):
        prior = False
        prior_tag = False
        chunk_dict, chunk_id = dict(), 0
        for i, tag in enumerate(tags):
            bio = tag[-1]
            if bio == "B":
                if prior:
                    chunk_dict[chunk_id]['end'] = i - 1
                chunk_id += 1
                chunk_dict[chunk_id] = {"entity_type": tag[:-2], "begin": i, "end": i}
                prior = True
                prior_tag = chunk_dict[chunk_id]['entity_type']  # 'SYMP'
                #print(prior_tag)
            elif bio == "I":
                if (prior_tag == False) or (tag[:-2] != prior_tag) : # 앞전의 태그가 O였거나 다른 태그였던 경우
                    #print(prior_tag, bio)
                    chunk_id +=1
                    chunk_dict[chunk_id] = {"entity_type": tag[:-2], "begin": i, "end": i}
                    prior = True
                    prior_tag =chunk_dict[chunk_id]['entity_type'] 
                else: # 앞전의 태그와 동일한 태그가 이어서 나온 경우
                    chunk_dict[chunk_id]['end'] = i 
            elif bio == "O":
                if prior:
                    chunk_dict[chunk_id]['end'] = i - 1
                prior = False
                prior_tag = False
        return chunk_dict

# to relation foramt

In [45]:
def to_relation_format(sent_list, tag_list):
        result = []
        for sent, tags in zip(sent_list, tag_list):
            
            #print(sent,tags)
            chunk_dict = merge_tags(tags)
            #print(chunk_dict)
            if len(chunk_dict) < 2:  # 엔티티 갯수가 2개 이상일 때만 포함
                continue
            for subj, obj in combinations(chunk_dict, 2):
                #print(chunk_dict[subj], chunk_dict[obj])
                re_sent = sent.copy()
                subj, obj = chunk_dict[subj], chunk_dict[obj]
                subj_begin, subj_end, obj_begin, obj_end = subj['begin'], subj['end'], obj['begin'], obj['end']
                re_sent[subj_begin] = "<e1>" + re_sent[subj_begin]
                re_sent[subj_end] = re_sent[subj_end] + "</e1>"
                re_sent[obj_begin] = "<e2>" + re_sent[obj_begin]
                re_sent[obj_end] = re_sent[obj_end] + "</e2>"
                result.append((" ".join(re_sent), subj["entity_type"], obj["entity_type"]))
        return result

In [49]:
output=to_relation_format(new_sent_list, new_tag_list)
output

[('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> <e2>부기</e2> ▲발열, 메스꺼움, 근육통 ▲피로감 ▲두통 등을 안내하고 있으나 생리불순이나',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> 부기 <e2>▲발열,</e2> 메스꺼움, 근육통 ▲피로감 ▲두통 등을 안내하고 있으나 생리불순이나',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> 부기 ▲발열, <e2>메스꺼움,</e2> 근육통 ▲피로감 ▲두통 등을 안내하고 있으나 생리불순이나',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> 부기 ▲발열, 메스꺼움, <e2>근육통</e2> ▲피로감 ▲두통 등을 안내하고 있으나 생리불순이나',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> 부기 ▲발열, 메스꺼움, 근육통 <e2>▲피로감</e2> ▲두통 등을 안내하고 있으나 생리불순이나',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> 부기 ▲발열, 메스꺼움, 근육통 ▲피로감 <e2>▲두통</e2> 등을 안내하고 있으나 생리불순이나',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 <e1>통증,</e1> 부기 ▲발열, 메스꺼움, 근육통 ▲피로감 ▲두통 등을 안내하고 있으나 <e2>생리불순이나</e2>',
  'SYMP',
  'SYMP'),
 ('이런 가운데 질병관리청은 예방 접종 후 이상 반응으로 ▲접종부위 통증, <e1>부기</e1> <e2>▲발열,</e2> 메스꺼움, 근육통 ▲피로감 ▲두통 등을